### PR TITLE
Add comirec Model.

### PR DESCRIPTION
**Pull request by RyanLee-Dev**
Main Change: Add comirec Model based on Mind.
Comirec has two version, self-attentive and  dynamic routing. The dynamic routing is almost same as Mind.

### CODE REVIEW

File: deepmatch/models/comirec.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        high_capsule = CapsuleLayer(input_units=item_embedding_dim,
+                                    out_units=item_embedding_dim, max_len=seq_max_len,
+                                    k_max=interest_num)((history_emb, hist_len))
+    elif interest_extractor.lower()=='sa':
+        history_emb_add_pos = history_emb
+        if add_pos:
+            position_embedding = PositionEncoding()(history_emb)
+            history_emb_add_pos += position_embedding # [None, MAX_LEN, emb_dim]
+
+        attn = DNN((item_embedding_dim*4, interest_num), dnn_activation, l2_reg_dnn,
```
Comment (shenweichen) [inline deepmatch/models/comirec.py:123]: why multiply 4 ?

File: deepmatch/models/comirec.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                        dnn_dropout, dnn_use_bn, output_activation=output_activation, seed=seed,
+                        name="user_dnn_attn")(history_emb_add_pos)
+        attn = tf.transpose(attn, [0, 2, 1]) # [None, interest_num, MAX_LEN]
+        seq_len_tile = tf.tile(hist_len, [1, interest_num])
+        mask = tf.sequence_mask(seq_len_tile, seq_max_len)  # [None, interest_num, MAX_LEN]
+        high_capsule = SoftmaxWeightedSum()([attn, history_emb_add_pos, mask])
```
Comment (shenweichen) [inline deepmatch/models/comirec.py:129]: `SoftmaxWeightedSum` has a default parameter `dropout_rate=0.2`,do we need it here?

File: deepmatch/models/comirec.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    high_capsule =None
+    if interest_extractor.lower()=='dr':
+        high_capsule = CapsuleLayer(input_units=item_embedding_dim,
+                                    out_units=item_embedding_dim, max_len=seq_max_len,
+                                    k_max=interest_num)((history_emb, hist_len))
+    elif interest_extractor.lower()=='sa':
+        history_emb_add_pos = history_emb
+        if add_pos:
+            position_embedding = PositionEncoding()(history_emb)
+            history_emb_add_pos += position_embedding # [None, MAX_LEN, emb_dim]
```
Comment (shenweichen) [inline deepmatch/models/comirec.py:121]: use `deepctr.layers.utils.add_func` instead of `+`

File: deepmatch/models/comirec.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+        attn = DNN((item_embedding_dim*4, interest_num), dnn_activation, l2_reg_dnn,
+                        dnn_dropout, dnn_use_bn, output_activation=output_activation, seed=seed,
+                        name="user_dnn_attn")(history_emb_add_pos)
+        attn = tf.transpose(attn, [0, 2, 1]) # [None, interest_num, MAX_LEN]
+        seq_len_tile = tf.tile(hist_len, [1, interest_num])
+        mask = tf.sequence_mask(seq_len_tile, seq_max_len)  # [None, interest_num, MAX_LEN]
```
Comment (shenweichen) [inline deepmatch/models/comirec.py:128]: try using a `Lambda` layers to encapsule these ops

**Review by shenweichen — CHANGES_REQUESTED** [review]
please follow the reviews to change codes

File: deepmatch/models/comirec.py (hunk 1 above)
Comment (liyuan97) [reply deepmatch/models/comirec.py:123]: Use the parameter or author's code.
https://github.com/THUDM/ComiRec/blob/master/src/model.py#L229

File: deepmatch/models/comirec.py (hunk 3 above)
Comment (liyuan97) [reply deepmatch/models/comirec.py:121]: Get.

File: deepmatch/models/comirec.py (hunk 2 above)
Comment (liyuan97) [reply deepmatch/models/comirec.py:129]: We don't. I will remove it.

File: deepmatch/models/comirec.py (hunk 4 above)
Comment (liyuan97) [reply deepmatch/models/comirec.py:128]: Got it.

**Pull request merged**